openzwavemqtt/util/mqtt_client.py

```diff
@@ -0,0 +1,246 @@
+        self,
+        topic: str,
+        qos: int = 0,
+        options: Optional[SubscribeOptions] = None,
+        properties: Optional[Properties] = None,
+        timeout: float = 10.0,
+    ) -> None:
+        """Subscribe to topic.
+
+        Can raise asyncio_mqtt.MqttError.
+        """
+        params: dict = {"qos": qos, "timeout": timeout}
+        if options:
+            params["options"] = options
+        if properties:
+            params["properties"] = properties
+
+        await self.asyncio_client.subscribe(topic, **params)
+
+    async def unsubscribe(
+        self, topic: str, properties: Optional[Properties] = None, timeout: float = 10.0
+    ) -> None:
+        """Unsubscribe from topic.
+
+        Can raise asyncio_mqtt.MqttError.
+        """
+        params: dict = {"timeout": timeout}
+        if properties:
+            params["properties"] = properties
+
+        await self.asyncio_client.unsubscribe(topic, **params)
+
+    def send_message(self, topic: str, payload: Union[str, dict]) -> None:
+        """Send a message from the manager options."""
+        to_publish = (topic, json.dumps(payload))
+        self.publish_queue.put_nowait(to_publish)
+
+    async def handle_publish(self) -> None:
+        """Publish messages as they are put on the queue."""
+        while True:
+            to_publish: tuple = await self.publish_queue.get()
+            await self.publish(*to_publish)
+            self.publish_queue.task_done()
+
+    async def start_client(self, manager: OZWManager) -> None:
+        """Start the client with the manager."""
+        self.client_task = asyncio.current_task()
+        # Reconnect automatically until the client is stopped.
+        while True:
+            try:
+                await self.subscribe_manager(manager)
+            except MqttError as err:
+                self.reconnect_interval = min(self.reconnect_interval * 2, 900)
+                LOGGER.error(
+                    "MQTT error: %s. Reconnecting in %s seconds",
+                    err,
+                    self.reconnect_interval,
+                )
+                await asyncio.sleep(self.reconnect_interval)
+                self.create_client()  # reset connect/reconnect futures
+
+    async def subscribe_manager(self, manager: OZWManager) -> None:
+        """Connect and subscribe to manager topics."""
+        async with AsyncExitStack() as stack:
+            # Keep track of the asyncio tasks that we create, so that
+            # we can cancel them on exit.
+            tasks: Set[asyncio.Task] = set()
+            stack.push_async_callback(cancel_tasks, tasks)
+
+            # Connect to the MQTT broker.
+            await stack.enter_async_context(self.asyncio_client)
+            # Reset the reconnect interval after successful connection.
+            self.reconnect_interval = 1
+
+            publish_task = asyncio.create_task(self.handle_publish())
+            tasks.add(publish_task)
```

Comment:
I would make `subscribe_manager` an internal method so it's clear people shouldn't pass in a 2nd, because right now we would launch 2 publish tasks.

Is there a reason publishing is a task with a queue and you don't just call `create_task(client.publish(…))` ?

Reply:
Good to make `subscribe_manager` private. :+1: 

By using a queue we keep the task management inside `subscribe_manager`. We want to cancel all sub tasks when we cancel the client task.